compiler/rustc_middle/src/ty/inhabitedness/inhabited_predicate.rs

```diff
@@ -23,42 +23,62 @@ pub enum InhabitedPredicate<'tcx> {
     /// Inhabited if some generic type is inhabited.
     /// These are replaced by calling [`Self::instantiate`].
     GenericType(Ty<'tcx>),
+    /// Inhabited if either we don't know the hidden type or we know it and it is inhabited.
+    OpaqueType(OpaqueTypeKey<'tcx>),
     /// A AND B
     And(&'tcx [InhabitedPredicate<'tcx>; 2]),
     /// A OR B
     Or(&'tcx [InhabitedPredicate<'tcx>; 2]),
 }
 
 impl<'tcx> InhabitedPredicate<'tcx> {
-    /// Returns true if the corresponding type is inhabited in the given
-    /// `ParamEnv` and module
+    /// Returns true if the corresponding type is inhabited in the given `ParamEnv` and module.
     pub fn apply(self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>, module_def_id: DefId) -> bool {
-        let Ok(result) = self.apply_inner::<!>(tcx, param_env, &mut Default::default(), &|id| {
-            Ok(tcx.is_descendant_of(module_def_id, id))
-        });
+        self.apply_reveal_opaque(tcx, param_env, module_def_id, &|_| None)
+    }
+
+    /// Returns true if the corresponding type is inhabited in the given `ParamEnv` and module,
+    /// revealing opaques when possible.
+    pub fn apply_reveal_opaque(
```

Comment:
Tiny nit, but could we call this `apply_revealing_opaque`? That matches other "do X, while also doing Y" methods like `normalize_erasing_regions`.

Reply:
Oh yeah that's better